src/compiler/sys.ts

```diff
@@ -126,6 +131,64 @@ namespace ts {
         }
     }
 
+    interface WatchedFileWithIsClosed extends WatchedFile {
+        isClosed?: boolean;
+    }
+    function pollWatchedFileQueue<T extends WatchedFileWithIsClosed>(
+        host: { getModifiedTime: NonNullable<System["getModifiedTime"]>; },
+        queue: (T | undefined)[],
+        pollIndex: number, chunkSize: number,
+        callbackOnWatchFileStat?: (watchedFile: T, pollIndex: number, fileChanged: boolean) => void
+    ) {
+        // Max visit would be all elements of the queue
+        let needsVisit = queue.length;
+        let definedValueCopyToIndex = pollIndex;
+        for (let polled = 0; polled < chunkSize && needsVisit > 0; nextPollIndex(), needsVisit--) {
+            const watchedFile = queue[pollIndex];
+            if (!watchedFile) {
+                continue;
+            }
+            else if (watchedFile.isClosed) {
+                queue[pollIndex] = undefined;
+                continue;
+            }
+
+            polled++;
+            const fileChanged = onWatchedFileStat(watchedFile, getModifiedTime(host, watchedFile.fileName));
+            if (watchedFile.isClosed) {
+                // Closed watcher as part of callback
+                queue[pollIndex] = undefined;
+            }
+            else {
+                callbackOnWatchFileStat?.(watchedFile, pollIndex, fileChanged);
+            }
+
+            if (queue[pollIndex]) {
+                // Copy this file to the non hole location

```

Comment:
Looks like a "move" if we're clearing the old location?  And what's the "non hole location"?

Reply:
Seems like maybe we're trying to defragment the queue?

Reply:
Yes we are defragmenting

Reply:
Again this is existing thing i just pulled it out so it can be shared for new option